src/main/java/com/amazonaws/encryptionsdk/kms/KmsMasterKeyProviderBuilder.java

```diff
@@ -0,0 +1,125 @@
+package com.amazonaws.encryptionsdk.kms;
+
+import com.amazonaws.ClientConfigurationFactory;
+import com.amazonaws.client.AwsSyncClientParams;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.regions.AwsRegionProvider;
+import com.amazonaws.regions.DefaultAwsRegionProviderChain;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.RegionUtils;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.kms.AWSKMSClient;
+import com.amazonaws.services.kms.AWSKMSClientBuilder;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Used to build a {@link KmsMasterKeyProvider} which will use the defaults found in the relevant provider chains.
+ *
+ * If a region is not defined and no region can be found in the region provider chain, then
+ * {@link Regions#DEFAULT_REGION} is used.
+ * <pre>
+ *     KmsMasterKeyProvider keyProvider = KmsMasterKeyProviderBuilder.defaultProvider();
+ * </pre>
+ *
+ * Example, specifying a region:
+ * <pre>
+ *     KmsMasterKeyProvider keyProvider = KmsMasterKeyProviderBuilder.standard()
+ *       .withRegion("us-east-1")
+ *       .build();
+ * </pre>
+ */
+public class KmsMasterKeyProviderBuilder extends AwsClientBuilder<KmsMasterKeyProviderBuilder, KmsMasterKeyProvider> {
+
+    private List<String> keyIds;
+
+    /**
+     * @return a new {@link KmsMasterKeyProviderBuilder} with the defaults set.
+     */
+    public static KmsMasterKeyProviderBuilder standard() {
+        return new KmsMasterKeyProviderBuilder();
+    }
+
+    /**
+     * @return a KmsMasterKeyProvider using the {@link AWSKMSClientBuilder} defaults, and an empty key ID list.
+     */
+    public static KmsMasterKeyProvider defaultProvider() {
+        return standard().build();
+    }
+
+    private KmsMasterKeyProviderBuilder() {
+        super(new ClientConfigurationFactory());
+    }
+
+    /**
+     * Sets the region to be used by the client. Overrides any previously set region.
+     */
+    public KmsMasterKeyProviderBuilder withRegion(Region region) {
+        return withRegion(region.getName());
+    }
+
+    /**
+     * Sets the region using a keyId. This will override any previously set regions.
```

Comment:
We should only infer the region from a key ID if the user hasn't manually configured a region, to avoid confusion. The rule is that we'd use `withDefaultRegion` first, followed by the implied region in `withKeyId`, followed by the SDK default region.

Reply:
This sounds good to me. Should we throw an exception if the user tries to use `withKeyId` after they have already set a default region? Or just continue with the region set with `withDefaultRegion`?

I'm leaning towards being explicit, and throwing an exception so that there is no unexpected behavior on the user's end. But I could be swayed.